v2/writer_test.go

```diff
@@ -131,3 +131,150 @@ func assertAddNodes(t *testing.T, adder format.NodeAdder, nds ...format.Node) {
+		wantErr bool
+	}{
+		{
+			name:    "CorruptPragmaIsRejected",
+			path:    "testdata/sample-corrupt-pragma.car",
+			wantErr: true,
+		},
+		{
+			name:    "CARv42IsRejected",
+			path:    "testdata/sample-rootless-v42.car",
+			wantErr: true,
+		},
+		{
+			name:    "CARv1RootsOfDifferentSizeAreNotReplaced",
+			path:    "testdata/sample-v1.car",
+			wantErr: true,
+		},
+		{
+			name:    "CARv2RootsOfDifferentSizeAreNotReplaced",
+			path:    "testdata/sample-wrapped-v2.car",
+			wantErr: true,
+		},
+		{
+			name:    "CARv1NonEmptyRootsOfDifferentSizeAreNotReplaced",
+			path:    "testdata/sample-v1.car",
+			roots:   []cid.Cid{merkledag.NewRawNode([]byte("QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n")).Cid()},
+			wantErr: true,
+		},
+		{
+			name:    "CARv1ZeroLenNonEmptyRootsOfDifferentSizeAreNotReplaced",
+			path:    "testdata/sample-v1-with-zero-len-section.car",
+			roots:   []cid.Cid{merkledag.NewRawNode([]byte("fish")).Cid()},
+			wantErr: true,
+		},
+		{
+			name:    "CARv2NonEmptyRootsOfDifferentSizeAreNotReplaced",
+			path:    "testdata/sample-wrapped-v2.car",
+			roots:   []cid.Cid{merkledag.NewRawNode([]byte("fish")).Cid()},
+			wantErr: true,
+		},
+		{
+			name:    "CARv2IndexlessNonEmptyRootsOfDifferentSizeAreNotReplaced",
+			path:    "testdata/sample-v2-indexless.car",
+			roots:   []cid.Cid{merkledag.NewRawNode([]byte("fish")).Cid()},
+			wantErr: true,
+		},
+		{
+			name:  "CARv1SameSizeRootsAreReplaced",
+			path:  "testdata/sample-v1.car",
+			roots: []cid.Cid{requireDecodedCid(t, "bafy2bzaced4ueelaegfs5fqu4tzsh6ywbbpfk3cxppupmxfdhbpbhzawfw5od")},
+		},
+		{
+			name:  "CARv2SameSizeRootsAreReplaced",
+			path:  "testdata/sample-wrapped-v2.car",
+			roots: []cid.Cid{requireDecodedCid(t, "bafy2bzaced4ueelaegfs5fqu4tzsh6ywbbpfk3cxppupmxfdhbpbhzawfw5oi")},
+		},
+		{
+			name:  "CARv2IndexlessSameSizeRootsAreReplaced",
+			path:  "testdata/sample-v2-indexless.car",
+			roots: []cid.Cid{requireDecodedCid(t, "bafy2bzaced4ueelaegfs5fqu4tzsh6ywbbpfk3cxppupmxfdhbpbhzawfw5oi")},
+		},
+		{
+			name:  "CARv1ZeroLenSameSizeRootsAreReplaced",
+			path:  "testdata/sample-v1-with-zero-len-section.car",
+			roots: []cid.Cid{requireDecodedCid(t, "bafy2bzaced4ueelaegfs5fqu4tzsh6ywbbpfk3cxppupmxfdhbpbhzawfw5o5")},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Make a copy of input files to preserve original for comparison.
+			// This also avoids modification files in testdata.
+			tmpCopy := requireTmpCopy(t, tt.path)
+			err := ReplaceRootsInFile(tmpCopy, tt.roots)
+			if tt.wantErr {
+				require.Error(t, err, "ReplaceRootsInFile() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			} else {
+				require.NoError(t, err)
+
+				original, err := os.Open(tt.path)
+				require.NoError(t, err)
+				t.Cleanup(func() { require.NoError(t, original.Close()) })
+
+				target, err := os.Open(tmpCopy)
+				require.NoError(t, err)
+				t.Cleanup(func() { require.NoError(t, target.Close()) })
```

Comment:
might as well just defer :) IMO Cleanup is only useful in test helper funcs, where defer can't be used.